### ISSUE TITLE
Text proposal for BERT-related issues

### ISSUE THREAD

**Issue by hannestschofenig** [core-wg/coap-tcp-tls#186]
Here is an initial text proposal for addressing various issues raised with BERT. Credits go to Oleg. 

--

The term "payload", as defined in RFC 7959, will be used for the actual content of a single CoAP message. The term "body" is used for the entire resource representation that is being transferred in a block-wise fashion. When blockwise transfer is used then the body will conveyed in the payloads of multiple CoAP messages.

The value in Max-Message-Size option specifies the maximum size of the payload in bytes that a sender of the Max-Message-Size option is willing to receive. Note that this value cannot directly be used as the maximum buffer size in a CoAP over TCP implementation since the payload itself does not include 
* the maximum length of the CoAP header (i.e., 14 bytes -- 2 byte header + 4 bytes extended len + 8 bytes token), 
* the maximum size of the options, 
* and the payload marker. 

Without allocating enough buffer space for these additional, variable length header fields there is the risk that the buffer will overflow. 

A sender MUST NOT advertise a value below 32 bytes in the Max-Message-Size option since otherwise the transfer of data will be stalled. The RECOMMENDED minimum value for the Max-Message-Size value is 1152 bytes. 

When the Max-Message-Size is set to a value that is less than 2048 then the use of BERT does not provide any benefits beyond the RFC 7959-defined blockwise transfer mechanism since SZX(6) with 1024 bytes will be used as a blocksize. 

A developer needs to decide what Max-Message-Size value to select, which should be guided by the available RAM of the device. If sufficient RAM is available, then a large value for Max-Message-Size should be selected since it will accellerate the transfer of data due to avoiding unnecessary chunking of data. 

When using TLS the Maximum Fragment Length (MFL) feature {{RFC6066}}, which is currently being revised with the Record Size Limit (RSL) Extension  {{I-D.ietf-tls-record-limit}}, allows to indicate the maximum size of protected records that each peer will send the other. Ideally, the value used for MFL/RSL should be aligned with the value provided in the Max-Message-Size option since otherwise the TLS layer reject the transmission of the data. 

--

**Comment by hannestschofenig** [comment]
Note: if we change the definition of Max-Message-Size  to [options + packet  marker + payload] then we obviously have to adjust the description. The advantage of this description is the simplicity; the disadvantage is that it is basically silent on the tough part (namely the unknown length of the options).

**Comment by hannestschofenig** [comment]
To add a remark to the interaction between Max-Message-Size (MMS) and the Record Size Limit (RSL) Extension. 

For TLS 1.3 the RSL value is defined to include application content + TLS content type + padding 

For TLS 1.2 the RSL value is defined to includes TLS content type + protocol version + length + fragment (=application content)

The application content consists of the CoAP header + options + packet marker + payload

When an application developer configures the MMS value then it has to set the RSL value in such a way that it is larger than MMS by taking the additional fields into account. As such, RSL has to be configured as 

RSL = MMS + CoAP header + maximum size of options + packet marker + TLS content type + padding (for TLS 1.3)
RSL = MMS + CoAP header + maximum size of options + packet marker + TLS content type + protocol version + length (for TLS 1.2)

If the value of RSS is set to a value that is smaller then the above-provided formula indicates then there is the risk that the TLS layer will reject the transmission of the packet. Unfortunately for developers some of the fields are of variable length, such as the options. The padding field in the TLSInnerPlaintext structure, which is included in the RSS value, is of variable length but will most likely be determined by the application. In addition to this padding, the cipher may itself require some padding for alignment with the blocksize the cryptographic algorithm operates with. This blockcipher padding is, however, excluded from the RSS calculation.

Note also that the way how RSL is defined in I-D.ietf-tls-record-limit does not take the additional overhead of the ciphers into account. An developer must therefore make the buffers larger than the value indicated in RSL since Authentication Encryption with Additional Data (AEAD) ciphers API requires that an entire record be present in RAM to decrypt and authenticate it.

**Comment by cabo** [comment]
On Aug 31, 2017, at 13:19, Hannes Tschofenig <notifications@github.com> wrote:
> 
> When an application developer configures the MMS value then it has to set the RSL value in such a way that it is larger than MMS 

The boundaries of messages are in no way related to the boundaries of TLS records.

Grüße, Carsten

**Comment by hannestschofenig** [comment]
Carsten, this is not my understanding. When TLS creates a record for transmission then it puts the plaintext into memory and applies the AEAD cipher to it. This will encrypt the payload and add an authentication tag to the end. The receiver has to then put this entire block into memory before applying the reverse operation.

Note that stream ciphers operate differently but their use has been deprecated with TLS.

**Comment by hannestschofenig** [comment]
Carsten, could you actually explain what you mean by "The boundaries of messages are in no way related to the boundaries of TLS records."?

**Comment by cabo** [comment]
On Aug 31, 2017, at 14:18, Hannes Tschofenig <notifications@github.com> wrote:
> 
> Carsten, could you actually explain what you mean by "The boundaries of messages are in no way related to the boundaries of TLS records.”?

One CoAP message may go into multiple TLS records, and one TLS record may contain (parts of) multiple CoAP messages (the interface between CoAP and TLS is a byte stream, where neither message boundaries are visible).  

Therefore, the TSL RSL is inconsequential to the CoAP message sizes and vice versa.

Grüße, Carsten

**Comment by hannestschofenig** [comment]
I dropped Martin Thomson a message to ask him about the anticipated relationship.

**Comment by hannestschofenig** [comment]
I got a response from Martin Thomson and here is the story. 

Without TLS, an application protocol using TCP can read a byte at a time out of the receive buffer (and control that by keeping the receive window small).  With TLS, you can not read a byte at a time because you can only read a record at a time due to the nature of the AEAD ciphers (as explained above already).

Hence, the Record Size Limit extension can help to select a proper size for the TLS fragments. So the text in the Record Size Limit extension that talks about "MUST NOT generate a protected record with plaintext that is larger than the RecordSizeLimit value" should rather be interpreted as a way to guide the size of the TLS fragments. 

When a developer wants to set the RSL and MSL to make them fit the available RAM size then they have to know how much overhead gets added (at maximum) throughout the stack to set the values appropriately.

The question for me is whether we should do some of that homework to tell the reader what aspects determine the overhead or not. Of course, being silent would be easier for us and setting the RSL value incorrectly can lead to two consequences: 

 * If the RSL value is not set or selected too large then the MSL value has essentially no effect since the TLS stack can still combine data and send larger payloads. 
 * If the RSL value is set too small then the TLS will fragment application data into smaller TLS fragments and the result is an inefficient transfer.

In a nutshell, I don't want to be the developer who has to decide what the appropriate value for RSL and MSL is given a specific hardware platform...

**Comment by cabo** [comment]
Re RSL: Is this the right place to give guidance to developers of TLS implementations?
This might have been a good discussion to add to RFC 7925.
Maybe we have a place in an LWIG document for this; e.g., lwig-coap.
Re MMS: I'm sure developer for constrained devices will be acutely aware of the issues here.
If we have something helpful to say here, let's put that into lwig-coap.

**Comment by hannestschofenig** [comment]
I spoke with Carsten about this topic and he believes that the CoAP over TCP is the wrong document to offer developer recommendations. He believes RFC 7925bis or an LWIG guidance document might be a better place.
